### DISCUSSION TITLE
allowedVersions ignored on dry run when repo is onboarding

### DISCUSSION THREAD

**Discussion by dhduvall** [renovatebot/renovate#22752] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

35.110.2

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

I'm setting up renovatebot for some Gitlab projects, and doing some testing on a local machine with the docker container. My execution is this:
```
docker run --rm -v $PWD/config.js:/usr/src/app/config.js --env-file env renovate/renovate:slim
```
The `env` file just contains credentials and `LOG_LEVEL=debug`.

<details><summary>My `config.js`:</summary>

```js
module.exports = {
    platform: "gitlab",
    repositories: [
        {
            repository: "dhduvall/test-renovate",
            baseBranches: ["duvall/renovateme"],
        },
    ],
    endpoint: process.env.CI_API_V4_URL,
    onboardingConfig: {
        extends: ["gitlab>dhduvall/renovate-config:default.json5#duvall/preset"],
    },
    onboardingConfigFileName: ".renovaterc.json5",

    gitAuthor: "Renovate Bot <group_...@noreply.gitlab.com>",

    hostRules: [ ]
```

</details>

<details><summary>The referenced custom preset is</summary>

```json5
{
    "$schema": "https://docs.renovatebot.com/renovate-schema.json",
    extends: [ "config:base" ],
    osvVulnerabilityAlerts: true,
    packageRules: [
        {
            description: ["package rule title"],
            matchManagers: ["maven"],
            // versions with qualifiers come before the unqualified version, but
            // renovate doesn't know that; we don't really want to suggest these
            // as an upgrade, anyway.
            // XXX Do we need .* before and after the -, or not? It seems like this is ignored on dry runs???
            allowedVersions: "!/-/",
        },
    ],
    // When the base branch is ahead of the renovate branch, the bot will always rebase.
    rebaseWhen: "behind-base-branch",
    // When a dependency is pulled, renovate will create a rollback MR.
    rollbackPrs: true,
}
```
</details>

<details>
<summary>When I run renovate as above with `--dry-run`, it tells me that it wants to upgrade a private maven dependency:</summary>

```json
               {
                 "datasource": "maven",
                 "depName": "net.coros:dd-dc-communication",
                 "currentValue": "1.0.69",
                 "fileReplacePosition": 1046,
                 "registryUrls": [
                   "https://repo.maven.apache.org/maven2",
                   "https://gitlab.com/api/v4/groups/14775601/-/packages/maven"
                 ],
                 "depType": "compile",
                 "groupName": "dd.dc.common.version",
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "1.0.69-duvall-gitlab-ci",
                     "newValue": "1.0.69-duvall-gitlab-ci",
                     "releaseTimestamp": "2023-05-06T20:23:04.000Z",
                     "newMajor": 1,
                     "newMinor": 0,
                     "updateType": "patch",
                     "branchName": "renovate/dd.dc.common.version"
                   }
                 ],
                 "packageName": "net.coros:dd-dc-communication",
                 "warnings": [],
                 "versioning": "maven",
                 "registryUrl": "https://gitlab.com/api/v4/groups/14775601/-/packages/maven",
                 "currentVersion": "1.0.69",
                 "isSingleVersion": true,
                 "fixedVersion": "1.0.69"
               },
```

</details>

which seems to contradict the `allowedVersions` settings I think it should be picking up.

<details><summary>However, if I drop `--dry-run`, it seems to pay attention to it:</summary>

```json
               {
                 "datasource": "maven",
                 "depName": "net.coros:dd-dc-communication",
                 "currentValue": "1.0.69",
                 "fileReplacePosition": 1046,
                 "registryUrls": [
                   "https://repo.maven.apache.org/maven2",
                   "https://gitlab.com/api/v4/groups/14775601/-/packages/maven"
                 ],
                 "depType": "compile",
                 "groupName": "dd.dc.common.version",
                 "updates": [],
                 "packageName": "net.coros:dd-dc-communication",
                 "warnings": [],
                 "versioning": "maven",
                 "registryUrl": "https://gitlab.com/api/v4/groups/14775601/-/packages/maven",
                 "currentVersion": "1.0.69",
                 "fixedVersion": "1.0.69"
               },
```
</details>

I can work on a minimal repro if it's needed, but before I spend a day on that I wanted to see if either the problem seemed familiar or if someone familiar with the codebase could reasonably easily figure it out from the source alone.

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction in the description above

**Comment by rarkins** [comment]
Dry run is no longer a boolean. Please set dryRun to extract, lookup, or full

I can't think of a reason why it would result in ignoring allowedVersions but you never know

**Reply by dhduvall** [reply]
The logs did contain
```
 WARN: cli config dryRun property has been changed to full
```
and
```
DEBUG: CLI config
       "config": {"dryRun": "full"}
```
so I chose `full` to match, and got the same behavior as before. `lookup` and `extract` show the same.

**Comment by rarkins** [comment]
Have you onboarded the repo yet? Or is this a repo still in onboarding phase?

**Reply by dhduvall** [reply]
Still in onboarding.

**Reply by rarkins** [reply]
Dry run plus an active onboarding is not really a supported concept because onboarding itself is essentially a dry run. What were you hoping to achieve?

**Reply by dhduvall** [reply]
I wanted to get the onboarding config itself correct, so that the `.renovaterc.json5` file had the contents I expected, and that those contents had the effect I wanted, as well as to explore exactly what the onboarding process looked like and what changes it would propose.

When exploring a new technology, I always like to have it tell me as much about what it's planning on doing before I actually let it do it so that I can catch all the surprises up front (especially if I'm not sure how to rewind the changes back to a pristine state to try again). This proposed upgrade was one of those surprises which I was trying to correct, when I ran into the surprise of dry run not matching what the actual operation proposed.

**Reply by rarkins** [reply, marked as answer]
Unfortunately this combination you are trying to use is not supported and likely complex to implement. You're welcome to investigate it further and submit a PR if you find otherwise

**Reply by dhduvall** [reply]
Okay; thanks!
